src/gretel_trainer/benchmark/session.py

```diff
@@ -0,0 +1,374 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Future, ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import Any, Optional, Union
+
+import pandas as pd
+from gretel_client.helpers import poll
+from gretel_client.projects import Project, create_project, search_projects
+from gretel_client.projects.jobs import Job
+
+from gretel_trainer.benchmark.core import BenchmarkConfig, BenchmarkException, Dataset
+from gretel_trainer.benchmark.custom.models import CustomModel
+from gretel_trainer.benchmark.custom.strategy import CustomStrategy
+from gretel_trainer.benchmark.executor import Executor
+from gretel_trainer.benchmark.gretel.models import GretelAuto, GretelModel
+from gretel_trainer.benchmark.gretel.strategy_sdk import GretelSDKStrategy
+from gretel_trainer.benchmark.gretel.strategy_trainer import GretelTrainerStrategy
+
+logger = logging.getLogger(__name__)
+
+ALL_REPORT_SCORES = {
+    "SQS": "synthetic_data_quality_score",
+    "FCS": "field_correlation_stability",
+    "PCS": "principal_component_stability",
+    "FDS": "field_distribution_stability",
+    "PPL": "privacy_protection_level",
+}
+
+
+class RunKey(tuple[str, str]):
+    RUN_IDENTIFIER_SEPARATOR = "-"
+
+    @property
+    def model_name(self) -> str:
+        return self[0]
+
+    @property
+    def dataset_name(self) -> str:
+        return self[1]
+
+    @property
+    def identifier(self) -> str:
+        """Identifier returns a flat string identifying this run.
+
+        Returns:
+            A string identifying the run.
+        """
+        return self.RUN_IDENTIFIER_SEPARATOR.join(self)
+
+    def __repr__(self) -> str:
+        return f"({self.model_name}, {self.dataset_name})"
+
+
+FutureKeyT = Union[str, RunKey]
+
+
+@dataclass
+class JobSpec:
+    # TODO(pm): for simplicity just use Dataset instead of union
+    dataset: Dataset
+
+    # TODO(pm): for simplicity, we only allow storing instantiated models here, no Type[X]
+    model: Union[GretelModel, CustomModel]
+
+
+def launch(*, jobs: list[JobSpec], config: Optional[BenchmarkConfig] = None) -> Session:
+    session = Session(jobs=jobs, config=config)
+    return session.prepare().execute()
```

Comment:
I definitely agree that `JobSpec` is a lot simpler when the types of the two fields are constrained/standardized to what you have defined here, and I like `Session` accepting a list of those objects to use internally with `compare` handling the process of creating that list. However, I think `launch` should accept a list of some different dataclass that, like `compare`, uses the wider types (`DatasetTypes`, `ModelTypes`).

Right now we document two kinds of datasets: Gretel-curated and create-your-own. I don't really want to have to explain "there's also a third, generic version of Datasets that both those two can be coerced to and that we really use under the hood...". In other words, I want to keep the API surface area smaller. Similarly, with models like `GretelACTGAN` (which is a class name, but in current usage kind of feels more like a constant), I think it'd be odd to have to explain "you can pass `GretelACTGAN` to `compare` but you can't use it as a model in a `JobSpec` passed to `launch`."

This implementation of `launch` doesn't really add a _whole_ lot over just manually instantiating a Session and calling prepare + execute. It feels to me like if someone wants to go down to a lower level, they should do exactly that (init+prepare+execute), but if they want to operate at a more convenient higher level they should use `compare` or `launch` and be able to pass nearly-identical arguments to those two functions. It kind of seems like instead of two "levels" of API abstraction (high = `compare/launch`, low = `Session(...).prepare().execute()`), there are three here, with `launch` being "somewhere in the middle" between high and low.

Reply:
> However, I think launch should accept a list of some different dataclass that, like compare, uses the wider types (DatasetTypes, ModelTypes).

I agree. For now I wanted to do "can I extract this thing and make tests pass" and another iteration would be for the cleanup, so thanks for all this feedback. That `launch` in the current version is just a placeholder. I agree that `launch` and `compare` should be symmetrical as much as possible.

The main reason why I didn't implement `launch` properly was a question of what it should accept. I think the best option would be to have more permissive `JobSpec`, but I wasn't sure about introducing another class just for that. Then it could be a list of tuples or 2 lists, which would have to be same length.

From the point of view of similarity with `compare`, I'd be leaning towards list of tuples I think. What do you think?

Reply:
OK yeah that "order of operations" on this branch totally makes sense, glad we're on the same page.

I think list of tuples is good. I definitely like that better than the two lists option. Compared to a new more permissive dataclass, my gut reaction was concern over "what if the user gets the order wrong", i.e. `(dataset, model)` vs. `(model, dataset)`, but thinking about it some more, if that happens I think it's fine to just raise whatever runtime error occurs, since we do all our "prep work" in advance before kicking off any jobs.